File: backend/app/api/v1/blog.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.models import BlogPost, LeadCapture, AuditLog
from app.schemas.schemas import BlogPostResponse, LeadCaptureCreate, LeadCaptureResponse, BlogPostCreate

router = APIRouter()

# Seed basic blog posts if empty
MOCK_POSTS = [
    {
        "title": "Mastering Harvard Referencing: A Step-by-Step Guide",
        "slug": "mastering-harvard-referencing",
        "content": "Accurate citations form the cornerstone of academic report writing. In this guide, we break down standard book, journal, and web formatting systems to elevate your citation structure...",
        "category": "Referencing Guides"
    },
    {
        "title": "Sleek Slide Layouts: Design Hacks for Presentation Design",
        "slug": "sleek-slide-layouts-design-hacks",
        "content": "Visual communication is key. Learn how to implement clear layout margins, limit text blocks, choose modern typography colors (like charcoal and teal), and design layouts that command attention.",
        "category": "Presentation Skills"
    },
    {
        "title": "Boost Your Focus: Time-Blocking for Academic Productivity",
        "slug": "time-blocking-academic-productivity",
        "content": "Discover how organizing study hours into dedicated 50-minute task blocks improves retention, mitigates exam stress, and helps you deliver academic reports on schedule.",
        "category": "Productivity"
    }
]


@router.get("", response_model=List[BlogPostResponse])
def get_posts(db: Session = Depends(get_db)):
    """Fetch all published educational study guides and tips articles."""
    posts = db.query(BlogPost).filter(BlogPost.published == True).all()
    if not posts:
        # Auto-seed mock data on first request
        for p in MOCK_POSTS:
            db_post = BlogPost(
                title=p["title"],
                slug=p["slug"],
                content=p["content"],
                category=p["category"],
                published=True
            )
            db.add(db_post)
        db.commit()
        posts = db.query(BlogPost).filter(BlogPost.published == True).all()
    return posts


@router.get("/{slug}", response_model=BlogPostResponse)
def get_post_by_slug(slug: str, db: Session = Depends(get_db)):
    """Fetch article details for a given slug."""
    post = db.query(BlogPost).filter(BlogPost.slug == slug, BlogPost.published == True).first()
    if not post:
        raise HTTPException(status_code=404, detail="Article not found")
    return post


@router.post("/lead", response_model=LeadCaptureResponse, status_code=status.HTTP_201_CREATED)
def submit_lead(lead_in: LeadCaptureCreate, db: Session = Depends(get_db)):
    """Capture newsletter subscriptions or free consult requests from public pages."""
    lead = LeadCapture(
        type=lead_in.type,
        email=lead_in.email,
        phone=lead_in.phone,
        name=lead_in.name,
        message=lead_in.message
    )
    db.add(lead)
    
    # Log lead creation audit
    log = AuditLog(
        action="lead_capture",
        details=f"Captured lead type {lead_in.type} for email: {lead_in.email}"
    )
    db.add(log)
    db.commit()
    db.refresh(lead)
    return lead

File: backend/src/routes/blog.ts
import { Router } from "express";
import { BlogPost, LeadCapture, AuditLog } from "../models/models.js";

const router = Router();

const MOCK_POSTS = [
  {
    title: "Mastering Harvard Referencing: A Step-by-Step Guide",
    slug: "mastering-harvard-referencing",
    content: "Accurate citations form the cornerstone of academic report writing. In this guide, we break down standard book, journal, and web formatting systems to elevate your citation structure...",
    category: "Referencing Guides"
  },
  {
    title: "Sleek Slide Layouts: Design Hacks for Presentation Design",
    slug: "sleek-slide-layouts-design-hacks",
    content: "Visual communication is key. Learn how to implement clear layout margins, limit text blocks, choose modern typography colors (like charcoal and teal), and design layouts that command attention.",
    category: "Presentation Skills"
  },
  {
    title: "Boost Your Focus: Time-Blocking for Academic Productivity",
    slug: "time-blocking-academic-productivity",
    content: "Discover how organizing study hours into dedicated 50-minute task blocks improves retention, mitigates exam stress, and helps you deliver academic reports on schedule.",
    category: "Productivity"
  }
];

router.get("", async (req, res) => {
  try {
    let posts = await BlogPost.findAll({ where: { published: true } });
    if (posts.length === 0) {
      for (const p of MOCK_POSTS) {
        await BlogPost.create({
          title: p.title,
          slug: p.slug,
          content: p.content,
          category: p.category,
          published: true
        });
      }
      posts = await BlogPost.findAll({ where: { published: true } });
    }
    return res.json(posts);
  } catch (err: any) {
    console.error(err);
    return res.status(500).json({ detail: err.message || "Failed to load articles" });
  }
});

router.get("/:slug", async (req, res) => {
  try {
    const post = await BlogPost.findOne({
      where: { slug: req.params.slug, published: true }
    });
    if (!post) {
      return res.status(404).json({ detail: "Article not found" });
    }
    return res.json(post);
  } catch (err: any) {
    console.error(err);
    return res.status(500).json({ detail: err.message || "Failed to load article details" });
  }
});

router.post("/lead", async (req, res) => {
  try {
    const { type, email, phone, name, message } = req.body;
    if (!type || !email) {
      return res.status(400).json({ detail: "Missing required lead fields" });
    }

    const lead = await LeadCapture.create({
      type,
      email,
      phone: phone || null,
      name: name || null,
      message: message || null
    });

    await AuditLog.create({
      action: "lead_capture",
      details: `Captured lead type ${type} for email: ${email}`
    });

    return res.status(201).json(lead);
  } catch (err: any) {
    console.error(err);
    return res.status(500).json({ detail: err.message || "Failed to capture lead" });
  }
});

export default router;
